### PR TITLE
Esther/gp 70/output group chat record

### PR DESCRIPTION
**Pull request by Esther125**
### Steps
Update the `.env` file
```
# LINE Bot Settings
CHANNEL_SECRET=[YOUR_CHANNEL_SECRET]
CHANNEL_ACCESS_TOKEN=[YOUR_CHANNEL_ACCESS_TOKEN]

# MongoDB Settings
MONGO_URI=mongodb://mongodb:27017
MONGO_DB_NAME=Graduation-Project
MONGO_COLLECTION=group-chat-record
```
Run the app
```
cd src\chatbot
python main.py
```
Follow the steps in the README.md to start ngrok and update _Webhook URL_ in LINE Developers settings. (**NOTICE:**  A correct _Webhook URL_ should look like this: `[YOUR_NGROK_URL]/linebot/callback`)

Create a group and add the official account into the group. Any message sent to the group will be saved to MongoDB. All the message records from the group will be logged to the console.
### Sample Result
  測試成功

### CODE REVIEW

**Comment by frafishen** [comment]
pls also update the `example.env`

**Comment by frafishen** [comment]
I'm not really sure if I missed some steps, cuz I got the error msgs below after adding the chatbot into the group.

```
INFO:     192.168.65.1:24623 - "POST /callback HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 411, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 69, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/srv/graduation-project/src/chatbot/main.py", line 48, in line_webhook
    and event["message"]["type"] == "text"  # noqa
KeyError: 'message'

```

File: src/chatbot/main.py [hunk 1]
```diff
@@ -9,31 +12,95 @@
...
+            await output_group_msg(group_id)
+        elif (
+            event["source"]["type"] == "user"
+            and event["message"]["type"] == "text"  # noqa
+        ):
+            await user_chat(event)
+        else:
+            print("The message is not in form of text.")
+
     return JSONResponse(status_code=200, content={"message": "OK"})
```
Comment (frafishen) [inline src/chatbot/main.py:64]: The functional functions and the routing functions should not be in the same file.

File: src/chatbot/main.py [hunk 2]
```diff
@@ -9,31 +12,95 @@
...
+            await database.add_message_to_group(group_id, msg_detail)
+            print(f"Insert message to {group_id} successfully.")
+        except Exception as e:
+            print(f"Exception: {e}")
+    return group_id
+
+
+async def user_chat(events):
+    pass
+
```
Comment (frafishen) [inline src/chatbot/main.py:94]: Also these two

**Comment by Esther125** [comment]
It seems that your event variable is out of format. Could you try to print out the events variable at line 43 in main.py? Thanks.

File: src/chatbot/main.py [hunk 3]
```diff
@@ -9,31 +12,95 @@
 from linebot.models import MessageEvent, TextMessage, TextSendMessage
 from linebot.v3.webhook import WebhookHandler
 
+import database
+
 app = FastAPI()
 
-# get environment variables
+
```
Comment (frafishen) [inline src/chatbot/main.py:19]: Consider prefixing routing paths for the Line chatbot to distinguish them clearly from the DB CRUD operations. 

add a line like
`router = APIRouter(prefix="/linebot")`

**Review by frafishen — CHANGES_REQUESTED** [review]

**Comment by frafishen** [comment]
> ld you try to print out the events variable at line 43 in main.py? Thanks.

quite normal???

------------------------------------------
[{'type': 'join', 'webhookEventId': '01HX67BBXNN485S272F5FQJMAX', 'deliveryContext': {'isRedelivery': False}, 'timestamp': 1714974731884, 'source': {'type': 'group', 'groupId': 'C37cf80e16fe3f373a1b6957dbe3870cb'}, 'replyToken': 'b271174fd3e9482e827ff157e0dbb3b9', 'mode': 'active'}]
------------------------------------------

**Comment by Esther125** [comment]
Normal message event looks like this:
```
EVENTS: [{'type': 'message', 'message': {'type': 'text', 'id': '506916745496756691', 'quoteToken': 'Ywa8ompgBK3auvkcRRBVUXqO_uy5AWRHBgexWDN7PxtcuRIyhY9BGc2Wrpkm48AdIfCuklxTEpRkqlUb2VyfLBqFbWaawjRRJb_LTHoczMTD61UkS9rZ-4JuaCCZ5xhk0sFugFrlcG8BRqM0WNR16g', 'text': '一樣'}, 'webhookEventId': '01HX69QK5P79J2K2HG7CPVX7QK', 'deliveryContext': {'isRedelivery': False}, 'timestamp': 1714977229501, 'source': {'type': 'group', 'groupId': 'Cdebbc45c673946cb3c85fb5ea318e0ab', 'userId': 'U55370f7ef8ef0a0b9a826a7e2e0fa5ba'}, 'replyToken': '2e1906ce73b34ae89b35b2b63c5b71d9', 'mode': 'active'}]
```
Your event is related to joining the group. You can just ignore it, restart the server, and start to send message to the group to test it. I will add a part to ignore joining event to fix the exception.

**Comment by yun-si** [comment]
I also got Error Message when adding the Line bot into a group.
```
INFO:     172.18.0.1:43830 - "POST /callback HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 411, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 69, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/srv/graduation-project/src/chatbot/main.py", line 48, in line_webhook
    and event["message"]["type"] == "text"  # noqa
KeyError: 'message'
Exception: object UpdateResult can't be used in 'await' expression
```

✨✨But the message record output is success.
```
----------------------------------------
FROM MONGO DB: {'_id': ObjectId('6638817eed161db61bb6a7e6'), 'groupId': '...', 'messages': [{'UserId': '...', 'UserName': '林耘熙', 'MsgText': 'hihi ', 'Time': datetime.datetime(2024, 5, 6, 7, 6, 38, 722000)}]}
```
ps. I don't know whether the id can be public, so I replace it to ... manually.

File: src/chatbot/main.py (hunk 1 above)
Comment (Esther125) [reply src/chatbot/main.py]: Done.

**Comment by frafishen** [comment]
Also works fine here!!!

**Review by frafishen — APPROVED** [review]

**Review by frafishen — CHANGES_REQUESTED** [review]
add the package of `service` into the `requirements.txt` in the root.
Rebuild the docker image, and the `uvicorn src.chatbot.main:app --host localhost --port 8080 --reload` can run well.
image

image

**Review by frafishen — APPROVED** [review]

**Comment by frafishen** [comment]
> add the package of `service` into the `requirements.txt` in the root. Rebuild the docker image, and the `uvicorn src.chatbot.main:app --host localhost --port 8080 --reload` can run well. 

I forgot we have a ticket already.
This PR is fine!!
https://frafishen.atlassian.net/browse/GP-83

**Pull request merged**